pdudaemon/drivers/conrad197720.py

```diff
@@ -0,0 +1,354 @@
+            if (data[(i>>2)+0] ^ data[(i>>2)+1] ^ data[(i>>2)+2]) != data[(i>>2)+3]:
+                print (f"checksum of frame {i} does not match")
+                return False
+        return True
+
+    def __sendCommand(self, cmd_byte, addr_byte, data_byte):
+        """
+        Private function to send a command to the card(s) / retry / parse the response
+        :param self: The object itself
+        :param cmd_byte: The command to the card(s)
+        :param addr_byte: The card address, used when more than one relay card is connected to the same serial port
+        :param data_byte: The data byte to be transmitted
+        :return: the response data byte
+        :rtype: int
+        :raises RuntimeError: all retries failed, communication error
+        """
+        for i in range (0,RETRY_COUNT):
+            self.__txFrame([cmd_byte,addr_byte,data_byte])
+            num_frames = 1
+            # for init we might receive n+1 frames in case there n cards cascaded
+            if cmd_byte==CMD_INIT:
+                num_frames = MAX_CARDS+1
+            recv_data = self.__rxFrame(num_frames)
+            # check if received data is valid
+            if len(recv_data) > 0 and True == self.__checkFrameChecksum(recv_data):
+                # retry, if there is no valid reply on init command
+                if cmd_byte==CMD_INIT:
+                    if (self.__checkNumCards(recv_data) <= 0):
+                        log.debug("__sendCommand retry reason 1")
+                        continue
+                # check if the addr and cmd in the reply is the correct answer
+                if recv_data[0]==255-cmd_byte and recv_data[1]==addr_byte:
+                    return recv_data[2]
+                # if not.. retry
+                else:
+                    log.debug("__sendCommand retry reason 2")
+                    continue
+            else:
+                # send up to 4 single bytes until we receive correct frames
+                # since there might be wrong communication before,
+                # especially in the beginning of communication
+                for j in range (0,4):
+                    self.__txSingleByte()
+                    recv_data = self.__rxFrame(255)
+                    if self.__checkFrameChecksum(recv_data):
+                        break
+                    log.debug("__sendCommand retry reason 3")
+        raise RuntimeError("All retries failed, communication error")
+
+    def __checkNumCards(self, data):
+        """
+        Private function to parse the amount of concatenated card at this serial port
+        :param self: The object itself
+        :return: number of cards
+        :rtype: int
+        """
+        self.num_cards=0
+        for i in range (0,(len(data)>>2)):
+            if (data[(i<<2)] == 254):
+                self.num_cards = self.num_cards + 1
+        log.debug(f"found {self.num_cards} cards")
+        return self.num_cards
+
+    def getNumCards(self):
```

Comment:
This and the following functions seem to be internal as well, so perhaps add the `__` prefix.

Reply:
You are right, I have added the prefix to all private functions.